src/runtime_src/core/edge/drm/zocl/scu.c

```diff
@@ -0,0 +1,267 @@
+
+static struct attribute *scu_attrs[] = {
+	&dev_attr_debug.attr,
+	&dev_attr_cu_stat.attr,
+	&dev_attr_cu_info.attr,
+	&dev_attr_stat.attr,
+	NULL,
+};
+
+static const struct attribute_group scu_attrgroup = {
+	.attrs = scu_attrs,
+};
+
+static int configure_soft_kernel(u32 cuidx, char kname[64], char uuid[16])
+{
+	struct drm_zocl_dev *zdev = zocl_get_zdev();
+	struct soft_krnl *sk = zdev->soft_kernel;
+	struct soft_krnl_cmd *scmd;
+	struct config_sk_image *cp;
+
+	cp = kmalloc(sizeof(struct config_sk_image), GFP_KERNEL);
+	cp->start_cuidx = cuidx;
+	cp->num_cus = 1;
+	strncpy((char *)cp->sk_name,kname,PS_KERNEL_NAME_LENGTH);
+	strcpy(cp->sk_uuid,uuid);
+
+	mutex_lock(&sk->sk_lock);
+
+	/*
+	 * Fill up a soft kernel command and add to soft
+	 * kernel command list
+	 */
+	scmd = kmalloc(sizeof(struct soft_krnl_cmd), GFP_KERNEL);
+	if (!scmd) {
+		DRM_WARN("Config Soft CU failed: no memory.\n");
+		mutex_unlock(&sk->sk_lock);
+		return -ENOMEM;
+	}
+
+	scmd->skc_opcode = ERT_SK_CONFIG;
+	scmd->skc_packet = cp;
+
+	list_add_tail(&scmd->skc_list, &sk->sk_cmd_list);
+	mutex_unlock(&sk->sk_lock);
+
+	/* start CU by waking up PS kernel handler */
+	wake_up_interruptible(&sk->sk_wait_queue);
+
+	return 0;
+}
+
+static int scu_probe(struct platform_device *pdev)
+{
+	struct zocl_scu *zcu;
+	struct xrt_cu_info *info;
+	struct drm_zocl_dev *zdev;
+	int err = 0;
+
+	zcu = kzalloc(sizeof(*zcu), GFP_KERNEL);
+	if (!zcu)
+		return -ENOMEM;
+
+	zcu->pdev = pdev;
+	zcu->base.dev = &pdev->dev;
+
+	info = dev_get_platdata(&pdev->dev);
+	memcpy(&zcu->base.info, info, sizeof(struct xrt_cu_info));
+
+	zdev = zocl_get_zdev();
+	err = zocl_kds_add_scu(zdev, &zcu->base);
+	if (err) {
+		DRM_ERROR("Not able to add SCU %p to KDS", zcu);
+		goto err;
+	}
+
+	err = xrt_cu_scu_init(&zcu->base);
+	if (err) {
+		DRM_ERROR("Not able to initial SCU %p\n", zcu);
+		goto err2;
+	}
+
+	platform_set_drvdata(pdev, zcu);
+
+	err = sysfs_create_group(&pdev->dev.kobj, &scu_attrgroup);
+	if (err)
+		zocl_err(&pdev->dev, "create SCU attrs failed: %d", err);
+
+	err = configure_soft_kernel(info->cu_idx,info->kname,info->uuid);
+	if (err)
+		zocl_err(&pdev->dev, "configuring SCU failed: %d", err);
+	
+	zocl_info(&pdev->dev, "SCU[%d] created", info->cu_idx);
+	return 0;
+err2:
+	zocl_kds_del_scu(zdev, &zcu->base);
+err:
+	kfree(zcu);
+	return err;
+}
+
+static int scu_remove(struct platform_device *pdev)
+{
+	struct zocl_scu *zcu;
+	struct drm_zocl_dev *zdev;
+	struct xrt_cu_info *info;
+
+	zcu = platform_get_drvdata(pdev);
+	if (!zcu)
+		return -EINVAL;
+
+	info = &zcu->base.info;
+	xrt_cu_scu_fini(&zcu->base);
+
+	zdev = zocl_get_zdev();
+	zocl_kds_del_cu(zdev, &zcu->base);
+
+	if (zcu->base.res)
+		vfree(zcu->base.res);
+
+	sysfs_remove_group(&pdev->dev.kobj, &scu_attrgroup);
+
+	zocl_info(&pdev->dev, "SCU[%d] removed", info->cu_idx);
+	kfree(zcu);
+
+	return 0;
+}
+
+static struct platform_device_id scu_id_table[] = {
+	{"SCU", 0 },
+	{ /* end of table */ },
```

Comment:
Hmm...How is this second time being used.  My guess is that if the first entry of the structure is empty, then the end of the array has been reached.  Is that correct?